trust: X.509 self issuance (#603)
italia/eidas-it-wallet-docs: docs/en/trust.rst

     - 
   * - Trusted List
     - Trust Anchor

>      - The listing endpoint, the trust mark status endpoint, and the fetch endpoint MUST be exposed by both Trust Anchors and Intermediates, making the Trusted List distributed over multiple Federation Entities, where each of these is responsible for their registered subordinates. Other endpoints using different data formats may be implemented to facilitate interoperability with systems that do not support OpenID Federation 1.0. In such cases, the same information about federation entities must be synchronized across these endpoints, ensuring consistent availability of information through different channels.

Author: ```suggestion
     - The listing endpoint, the trust mark status endpoint, and the fetch endpoint MUST be exposed by both Trust Anchors and Intermediates, making the Trusted List distributed over multiple Federation Entities, where each of these is responsible for their registered subordinates. Other endpoints using different data formats MAY be implemented to facilitate interoperability with systems that do not support OpenID Federation 1.0. In such cases, the same information about federation entities MUST be synchronized across these endpoints, ensuring consistent availability of information through different channels.
```

General Properties
------------------

> The architecture of the trust infrastructure is built upon several core principles:

Author: ```suggestion
The architecture of the trust infrastructure is built upon the following core principles:
```

     - While part of a federated ecosystem, each entity retains control over its own definitions and configurations.
   * - P9
     - **Decentralization**

>      - Unlike traditional centralized systems, the trust infrastructure should allow a decentralized approach. No single entity has control over the entire system.

Author: ```suggestion
     - Unlike traditional centralized systems, the trust infrastructure should allow a decentralized approach.
```

Even if the Trust Anchor has changed its cryptographic keys for digital signature, the Federation Historical Keys endpoint always makes the keys no longer used available for historical signature verifications.

X.509 PKI
---------

The X.509 Public Key Infrastructure (PKI) is a framework designed to create, manage, distribute, use, store, and revoke digital X.509 Certificates. At the heart of X.509 PKI is the concept of a Certificate Authority (CA), which issues digital certificates to entities. These certificates are required for establishing secure communications over networks, including the internet, by enabling encryption and digital signature functionalities. The PKI hierarchy typically involves a root CA at the top, with one or more subordinate CAs beneath, forming a trusted chain. Entities rely on this chain of trust to verify the authenticity of certificates. X.509 standards define the format of public key certificates.

The integration of OpenID Federation 1.0 with the traditional X.509 based PKI (rfc:5280), complemented by a RESTful API, aims to enhance the infrastructure with additional features, making it navigable and transparent.

> This approach leverages the dynamic and flexible nature of OpenID Federation alongside the requirement of the X.509 Certificates for legacy applications and interoperability purposes, aiming to addresses the evolving needs of verification of the registration status of the federation participants, their compliance to the shared rules and the general trust management in multilateral digital ecosystems aiming to be interoperable with each other.

Author: ```suggestion
This approach leverages the dynamic and flexible nature of OpenID Federation alongside the requirement of the X.509 Certificates for legacy applications and interoperability purposes, aiming to addresses the evolving needs of verification of the registration status of the federation participants, their compliance to the shared rules and the general and interoperable trust management in multilateral digital ecosystems.
```

OpenID Federation and X.509 based PKI share several things in common, as listed below:

- **Hierarchical Approach**: both utilize a hierarchical trust model with a single, overarching trusted third party, known as the Trust Anchor, which is trusted above all others.

> - **Decentralization with Multiple Trust Anchors and Intermediates**: despite a unique hierarchical model, the possibility of having multiple Trust Anchors and intermediates, below one or more Trust Anchors, introduces a level of decentralization.

Author: ```suggestion
- **Decentralization with Multiple Trust Anchors and Intermediates**: despite a unique hierarchical model, the possibility of having multiple Trust Anchors and Intermediates, below one or more Trust Anchors, introduces a level of decentralization.
```

> - **Custom Extensions**: both systems allow for custom extensions to meet specific requirements or to enhance functionality. X.509 Certificates support custom extensions, OpenID Federation allows definition of custom protocol specific metadata, trust marks and policies using a policy language.

Author: ```suggestion
- **Custom Extensions**: both systems allow for custom extensions to meet specific requirements or to enhance functionality. X.509 Certificates support custom extensions, OpenID Federation allows definition of custom protocol specific metadata, Trust Marks and policies using a policy language.
```

> - **Trust/Certificate Chain**: they rely on a chained proof of trust, where trust is passed down from the root authority (Trust Anchor) through intermediaries to the end entity (Leaf).

Author: ```suggestion
- **Trust/Certificate Chain**: they rely on a chained proof of trust, where trust is passed down from the root authority (Trust Anchor) through Intermediaries to the end entity (Leaf).
```

> - **Constraints in the Chain**: constraints can be applied within the Trust Chain regarding critical aspects such as the delegation of trust, the number of intermediates, and the domains involved.

Author: ```suggestion
- **Constraints in the Chain**: constraints can be applied within the Trust Chain regarding critical aspects such as the delegation of trust, the number of intermediaries, and the domains involved.
```

- **Public Key Distribution**: Both systems involve the distribution of the public key of the Trust Anchor to ensure entities can verify the trust chain.
- **Registry of Expired Keys**: Maintaining a registry of expired keys is crucial for both, ensuring non-repudiation of past signatures even when keys change.

Federation Trust Anchor and X.509 CA
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In the context of OpenID Federation, the Trust Anchor plays a role similar to that of a Certificate Authority (CA) in X.509-based Public Key Infrastructures (PKIs). Both serve as foundational elements of trust within their respective systems. In this document, the term "Trust Anchor" is often used to encompass both concepts. The trust infrastructure described here aligns the OpenID Federation Trust Anchor with the X.509 PKI Certificate Authority, making therefore them a single unique entity supporting both `RFC5280`_ and OpenID Federation 1.0.

X.509 Certificates Issuance
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In an OpenID Federation, each participant is required to self-issue its Entity Configuration, signing it with one of its cryptographic keys that are attested by its Immediate Superiors.

Author: ```suggestion
In an OpenID Federation, each participant is required to self-issue its Entity Configuration, signing it with one of its cryptographic keys that are attested by Immediate Superiors.
```

In the same way, each federation Entity has the autonomy to issue a signed statement about itself in the form of a X.509 Certificate.

> Federation participants that need to issue X.509 Certificates about themselves and for their specific purposes, can issue and sign X.509 Certificates using one of their Federation Entity Keys attested by their Federation Authorities (Immediate Superior). This process aligns the issuance of X.509 Certificates with the federation's delegation paradigm.

Reviewer: The statement around how 'can issue and sign X.509 Certificates using one of the Federation Entities Keys' to me feels like that entity key means it is a CA below that which is signs.  So, if one were to rely on the regular X509 tools, there could be room for problems like other technologies like eduroam have encountered which are that entity certificates are 'CA's but do not decorate it with the necessary booleans for cA -- see https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9 for section 4.2.1.9 basic constraints.  If what is written in the lines 758 are NOT in line with general x.509 practice, then the validation of the cert chain using regular tools will _not_ validate because the basic constraints will not be satisfied. It may be worthwhile clarifying non-leaf valid entity certificates and how to validate them using regular tooling.  Feedback welcomed.

Author: they are not CA, as shown in the non normative example (chain) CA is set to false.
They are intermediary constrained to issue certificates only about themselves.

Are you saying that intermediate entities must have the `X509v3 Basic Constraints: CA:TRUE" to pass the chain validation tests?

Author: my notes below

1. in many federated systems, intermediate certificates may not need to act as CAs.
2. however, you are saying that standard x.509 certificate chain tools might raise exception if all the intermediate certificates are not set with CA:TRUE, where only the leaf certificate must not have the constraint CA set to TRUE

if the point number 2 is confirmed, I can do this change to avoid exceptions. thank you @teamktown

Reviewer: @peppelinux , for item 2, it appears that it Conformant CAs will function more appropriately with existing tooling.

 Below is a minimal OpenSSL demo showing how a subordinate certificate must have CA:TRUE in BasicConstraints to function as a valid intermediate CA. If you set CA:FALSE for the intermediate and then try to use it to sign another certificate, path validation will fail.
```shell
#!/usr/bin/env bash

# 1) Create a self-signed root CA
openssl req -x509 -nodes -days 3650 \
  -newkey rsa:2048 \
  -subj "/CN=Root CA" \
  -keyout rootCA.key \
  -out rootCA.crt

# 2) Create the intermediate key and CSR
openssl genrsa -out inter.key 2048
openssl req -new -key inter.key \
  -subj "/CN=Intermediate" \
  -out inter.csr

# 3) Create an OpenSSL config for the intermediate (CA=TRUE or CA=FALSE)
cat > inter_ca_true.cnf <<EOF
[ext]
basicConstraints=critical, CA:TRUE, pathlen:0
keyUsage=critical, cRLSign, keyCertSign
EOF

# 4) Sign the intermediate with the root (CA=TRUE version)
openssl x509 -req -in inter.csr -CA rootCA.crt -CAkey rootCA.key -CAcreateserial \
  -days 365 -out inter_ca_true.crt -extfile inter_ca_true.cnf -extensions ext

# 5) Create a leaf key and CSR
openssl genrsa -out leaf.key 2048
openssl req -new -key leaf.key \
  -subj "/CN=Leaf Cert" \
  -out leaf.csr

# 6) Sign leaf with the intermediate (which has CA=TRUE)
openssl x509 -req -in leaf.csr -CA inter_ca_true.crt -CAkey inter.key -CAcreateserial \
  -days 365 -out leaf_from_ca_true.crt

# 7) Verify the chain (should succeed)
openssl verify -CAfile rootCA.crt -untrusted inter_ca_true.crt leaf_from_ca_true.crt

# 8) Repeat with an intermediate that sets CA=FALSE
cat > inter_ca_false.cnf <<EOF
[ext]
basicConstraints=critical, CA:FALSE
keyUsage=critical, cRLSign, keyCertSign
EOF

openssl x509 -req -in inter.csr -CA rootCA.crt -CAkey rootCA.key -CAcreateserial \
  -days 365 -out inter_ca_false.crt -extfile inter_ca_false.cnf -extensions ext

# Leaf signing attempt is the same, but now the intermediate has CA=FALSE
openssl x509 -req -in leaf.csr -CA inter_ca_false.crt -CAkey inter.key -days 365 \
  -out leaf_from_ca_false.crt 2>/dev/null

# 9) Verification fails: "error 20 at 0 depth lookup: unable to get local issuer certificate"
openssl verify -CAfile rootCA.crt -untrusted inter_ca_false.crt leaf_from_ca_false.crt

```
Sample run:
```Shell
./test.sh 
.+....+.....+......+...+....+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*...+...+.+..+...+.+...+..+..........+.....+.+...+.........+......+...........+....+..+....+.........+.........+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.+.....+.......+.....+...+...+....+......+.....+.+........+.............+.....+....+...+.....+...+........................+.+..+...+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
........+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.....+.+............+.................+.+.....+.........+...+.+.........+...+...+..+....+........+............+...+.+.....+...............+.+......+........+......+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*...+......+......+............................+.....+.........+.+...........+..........+...+..+...+...+.........+...............+...+.......+.........+.....+..........+.....+.....................+...+.+.........+..+.........+....+......+.....+............+....+..............+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
-----
Certificate request self-signature ok
subject=CN = Intermediate
Certificate request self-signature ok
subject=CN = Leaf Cert
leaf_from_ca_true.crt: OK
Certificate request self-signature ok
subject=CN = Intermediate
CN = Intermediate
error 79 at 1 depth lookup: invalid CA certificate
error leaf_from_ca_false.crt: verification failed
```

Author: I will definitively add CA:TRUE

this apprently matches the philosophy behind openid federation, where every participant can be a Trust Anchor about itself or other federations. This can be formally defined using X.509 certificate, using that constraint.

The result is that all the certificates about and from the trust anchor to the leaf will have CA:TRUE

> Federation participants that need to issue X.509 Certificates about themselves and for their specific purposes, can issue and sign X.509 Certificates using one of their Federation Entity Keys attested by their Federtion Authorities (Superior Entities). This process aligns the issuance of X.509 Certificates with the federation's delegation paradigm.

Author: ```suggestion
Federation participants that need to issue X.509 Certificates about themselves and for their specific purposes, can issue and sign X.509 Certificates using one of their Federation Entity Keys attested by their Federation Authorities (Immediate Superior). This process aligns the issuance of X.509 Certificates with the federation's delegation paradigm.
```

> This is feasible because the X.509 Certificate can be verified using a Trust Chain, similar to the approach used for Entity Configurations in OpenID Federation.

Author: ```suggestion
This is feasible because the X.509 Certificate can be verified using a X.509 Certificate Chain, similar to the approach used for Entity Configurations in OpenID Federation.
```

Federation Leaves are not Certificate Authorities (CAs) or CA intermediaries authorized to issue X.509 certificates for their subordinates. Instead, Federation Leaves act as intermediaries for issuing certificates solely about themselves. This is accomplished by applying appropriate naming constraints to ensure that X.509 certificates are correctly scoped.
Naming constraints are applied by Immediate Superiors within the certificates issued to the Leaf entity, specifically concerning the Leaf's Federation Entity Keys. As a result, the Leaf can only issue X.509 certificates about itself, thereby maintaining the integrity of the Trust Chain.

.. name-constraints::

>    :permitted: URI.1=http://leaf.example.com

Reviewer: should the URI.1 be https or encourage 'permitted' on it?

Author: sure.

Author: done

>    :permitted: URI.1=http://leaf.example.com

Author: ```suggestion
   :permitted: URI.1=https://leaf.example.com
```

> .. name-constraints::
>    :permitted: URI.1=https://leaf.example.com
>    :permitted: CN=leaf.example.com

Author: ```suggestion
.. name-constraints::
   :permitted: URI.1=https://leaf.example.com
   :permitted: CN=leaf.example.com
   :permitted: DNS=leaf.example.com
   :excluded: DNS=localhost
   :excluded: DNS=localhost.localdomain
   :excluded: DNS=127.0.0.1
   :excluded: DNS=example.com
   :excluded: DNS=example.org
   :excluded: DNS=example.net
   :excluded: DNS=*.example.org
```

> When a participant self-issues an X.509 Certificate, it adheres to the following requirements:

Reviewer: It may be worthwhile identifying if there are an embargoed/forbidden values here.  Ones that come to mind are: localhost, localhost.localdomain, localhost.domain, the IETF restricted example.com / example.org domains, 127.0.0.1 and what about names without dots? Are they valid? what about the strings that look like wildcards but are not? *.example.org -- are wildcards forbidden?  Is there an obligation that the certificate CN / SAN DNS must persist only at issuance or into perpetuity? Think of how infrequent the intermediates will 'roll' or may reference and if the DNS name is lost or overtaken by someone else, does THAT pose a risk to the trust chain?  Thoughts welcome.

Author: yeah, it would be something like the one below:

````
.. name-constraints::
   :permitted: URI.1=https://leaf.example.com
   :permitted: CN=leaf.example.com
   :permitted: DNS=leaf.example.com
   :excluded: DNS=localhost
   :excluded: DNS=localhost.localdomain
   :excluded: DNS=127.0.0.1
   :excluded: DNS=example.com
   :excluded: DNS=example.org
   :excluded: DNS=example.net
   :excluded: DNS=*.example.org
````

I want also to add that in this context, wildcards cannot be allowed for Leaf's CA, since the Leaf can only issue certificates about their entity id hostname and uri.

1. **Subject Name**: The X.509 Certificate's subject name MUST match the participant's identity. Specifically, the ``Common Name (CN)`` field should contain the Federation Entity unique identifier value, which corresponds to the **sub** (subject) value in its federation Entity Configuration.
2. **Subject Alternative Name (SAN)**: The X.509 Certificate MUST include a ``SAN URI`` that matches the **sub** value of its federation Entity Configuration.
3. **DNS Name**: The X.509 Certificate MUST include a DNS Name in the SAN that matches the DNS name contained within the **sub** value of its Entity Configuration.
4. **Certificate Revocation List (CRL)**: The participant MUST publish a CRL for its self-issued X.509 Certificates. This list MUST be accessible and regularly updated to ensure that any compromised or invalid X.509 Certificates are promptly revoked with the motivation of the revocation, if any.

Below a non-normative example of an X.509 Certificate Chain without intermediaries and in plain text, to facilitate the reading.

.. code-block::

    Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
    Signature Algorithm: sha256WithRSAEncryption

>         Issuer: CN=https://trust-anchor.example.com, O=Example Trust Anchor, C=IT

Author: ```suggestion
        Issuer: CN=trust-anchor.example.com, O=Example Trust Anchor, C=IT
```

        Validity
            Not Before: Sep 1 00:00:00 2023 GMT
            Not After : Sep 1 00:00:00 2033 GMT

>         Subject: CN=https://trust-anchor.example.com, O=Example Trust Anchor, C=IT

Author: ```suggestion
        Subject: CN=trust-anchor.example.com, O=Example Trust Anchor, C=IT
```

        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (4096 bit)
                Modulus:
                    00:af:82:3b:...
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: 
                CA:TRUE
            X509v3 Key Usage: 
                Certificate Sign, CRL Sign
            X509v3 Subject Key Identifier: 
                12:34:56:78:90:ab:cd:ef:12:34:56:78:90:ab:cd:ef
            X509v3 Authority Key Identifier: 
                keyid:12:34:56:78:90:ab:cd:ef:12:34:56:78:90:ab:cd:ef
            X509v3 CRL Distribution Points: 
                Full Name:
                  URI:https://trust-anchors.example.com/crl/ca.crl
                  
    Signature Algorithm: sha256WithRSAEncryption
         5c:4f:3b:...

    Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1234567890 (0x499602d2)
    Signature Algorithm: sha256WithRSAEncryption

>         Issuer: CN=https://trust-anchor.example.com, O=Example Trust Anchor, C=IT

Author: ```suggestion
        Issuer: CN=trust-anchor.example.com, O=Example Trust Anchor, C=IT
```

        Validity
            Not Before: Sep 1 00:00:00 2023 GMT
            Not After : Sep 1 00:00:00 2024 GMT

>         Subject: CN=https://leaf.example.org, O=Leaf, C=IT

Author: ```suggestion
        Subject: CN=leaf.example.org, O=Leaf, C=IT
```

        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:af:82:3b:...
                Exponent: 65537 (0x10001)
        X509v3 extensions:

>             X509v3 Basic Constraints: 
>                 CA:FALSE

Author: ```suggestion
            X509v3 Basic Constraints: 
                CA:TRUE
```

        Version: 3 (0x2)
        Serial Number: 987654321 (0x3ade68b1)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=https://leaf.example.org, O=Leaf, C=IT
        Validity
            Not Before: Sep 1 00:00:00 2023 GMT
            Not After : Sep 1 00:00:00 2024 GMT
        Subject: CN=https://leaf.example.org, O=Leaf, C=IT
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:af:82:3b:...
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Key Usage: 
                Digital Signature, Key Encipherment
            X509v3 Subject Alternative Name: 
                URI:https://leaf.example.org
            X509v3 Name Constraints: 
                Permitted:
                  URI.1=https://leaf.example.com
                  DNS.1=leaf.example.com
            X509v3 CRL Distribution Points: 
                Full Name:
                  URI:https://leaf.example.org/crl/self.crl

    Signature Algorithm: sha256WithRSAEncryption
         7d:6e:5f:...

> Federation participants can ensure that their certificates are consistent, enabling interoperability and security across the federation. This approach introduces innovative practices for certificate management using the trust relationships established within the OpenID Federation.

Author: ```suggestion
Federation participants can ensure that their certificates are consistent, enabling interoperability and security across the federation. This approach, enabling X.509 certificate issuance delegation, introduces innovative practices for certificate management using the underlying trust relationships established within the OpenID Federation.
```

X.509 Certificate Revocation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

An X.509 Certificate can be revoked by its Issuer.

> When the X.509 Certificate issuer is the Leaf and therefore the X.509 Certificate is about itself, it MUST update its CRL.

Reviewer 2: In the case of short-lived certificates such as those issued by the Wallet Provider, it is not necessary to have CRLs. See https://www.ietf.org/archive/id/draft-housley-lamps-norevavail-01.html

Author: CRL it should be not mandatory, regarding the possibility to have short lived certificates.
I suggest to have a MUST have CRL for long-lived attestations, such as those with an expiration time superior than 24hours.

Author: Done here https://github.com/italia/eid-wallet-it-docs/pull/603/files#diff-b00d0cc50af3f9e721a6b5a70f4939182119581219f886e85833e9a133a51101R922

> When the X.509 Certificate issuer is an Immediate superior, such as the Trust Anchor or a Intermediate, it revokes the certificate about the leaf, therefore the X.509 Certificate about one of the Leafs Federation Entity Key. This action invalidates the entire Trust Chain associated with that Leaf, effectively removing its ability to issue further X.509 Certificates about itself. This hierarchical revocation mechanism ensures that any compromise or misbehavior by a Leaf entity can be swiftly addressed.

Author: ```suggestion
When the X.509 Certificate issuer is an Immediate superior, such as the Trust Anchor or a Intermediate, it revokes the certificate about the leaf, therefore the X.509 Certificate about one of the Leaves Federation Entity Key. This action invalidates the entire Trust Chain associated with that Leaf's cryptographic public key, effectively removing its ability to issue further X.509 Certificates about itself. This hierarchical revocation mechanism ensures that any compromise or misbehavior by a Leaf entity can be swiftly addressed.
```

                CA:TRUE
            X509v3 Key Usage: 
                Certificate Sign, CRL Sign
            X509v3 Subject Key Identifier: 
                12:34:56:78:90:ab:cd:ef:12:34:56:78:90:ab:cd:ef
            X509v3 Authority Key Identifier: 
                keyid:12:34:56:78:90:ab:cd:ef:12:34:56:78:90:ab:cd:ef
            X509v3 CRL Distribution Points: 
                Full Name:
                  URI:https://trust-anchors.example.com/crl/ca.crl
                  
    Signature Algorithm: sha256WithRSAEncryption
         5c:4f:3b:...

    Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1234567890 (0x499602d2)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=trust-anchor.example.com, O=Example Trust Anchor, C=IT
        Validity
            Not Before: Sep 1 00:00:00 2023 GMT
            Not After : Sep 1 00:00:00 2024 GMT
        Subject: CN=leaf.example.org, O=Leaf, C=IT
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:af:82:3b:...
                Exponent: 65537 (0x10001)
        X509v3 extensions:

>             X509v3 Basic Constraints: 
>                 CA:TRUE

Author: ```suggestion
            X509v3 Basic Constraints: 
                CA:TRUE, pathlen:1
```

            X509v3 Key Usage: 
                Digital Signature, Key Encipherment
            X509v3 Subject Alternative Name: 
                URI:https://leaf.example.com

>             X509v3 Name Constraints: 
>                 Permitted:
>                   URI.1=https://leaf.example.com
>                   DNS.1=leaf.example.com

Author: ```suggestion
            X509v3 Name Constraints: 
                Permitted:
                  URI.1=https://leaf.example.com
                  DNS.1=leaf.example.com
                Excluded:
                  DNS=localhost
                  DNS=localhost.localdomain
                  DNS=127.0.0.1
                  DNS=example.com
                  DNS=example.org
                  DNS=example.net
                  DNS=*.example.org
```

Federation Leaves are not Certificate Authorities (CAs) or CA intermediaries authorized to issue X.509 certificates for their subordinates. Instead, Federation Leaves act as intermediaries for issuing certificates solely about themselves. This is accomplished by applying appropriate naming constraints to ensure that X.509 certificates are correctly scoped.
Naming constraints are applied by Immediate Superiors within the certificates issued to the Leaf entity, specifically concerning the Leaf's Federation Entity Keys. As a result, the Leaf can only issue X.509 certificates about itself, thereby maintaining the integrity of the Trust Chain.

.. name-constraints::
   :permitted: URI.1=https://leaf.example.com
   :permitted: CN=leaf.example.com
   :permitted: DNS=leaf.example.com
   :excluded: DNS=localhost
   :excluded: DNS=localhost.localdomain
   :excluded: DNS=127.0.0.1
   :excluded: DNS=example.com
   :excluded: DNS=example.org
   :excluded: DNS=example.net
   :excluded: DNS=*.example.org

When a participant self-issues an X.509 Certificate, it adheres to the following requirements:

> 1. **Subject Name**: The X.509 Certificate's subject name MUST match the participant's identity. Specifically, the ``Common Name (CN)`` field should contain the Federation Entity unique identifier value, which corresponds to the **sub** (subject) value in its federation Entity Configuration.
> 2. **Subject Alternative Name (SAN)**: The X.509 Certificate MUST include a ``SAN URI`` that matches the **sub** value of its federation Entity Configuration.
> 3. **DNS Name**: The X.509 Certificate MUST include a DNS Name in the SAN that matches the DNS name contained within the **sub** value of its Entity Configuration.
> 4. **Certificate Revocation List (CRL)**: The participant MUST publish a CRL for its self-issued X.509 Certificates. This list MUST be accessible and regularly updated to ensure that any compromised or invalid X.509 Certificates are promptly revoked with the motivation of the revocation, if any.

Author: ```suggestion
1. **Subject Name**: The X.509 Certificate's subject name MUST match the participant's identity. Specifically, the ``Common Name (CN)`` field should contain the Federation Entity unique identifier DNS name, which is included into the **sub** (subject) value in its federation Entity Configuration, removing ``https://`` and any webpaths.
2. **Subject Alternative Name (SAN)**: The X.509 Certificate MUST include a ``SAN URI`` that matches the **sub** value of its federation Entity Configuration.
3. **DNS Name**: The X.509 Certificate MUST include a DNS Name in the SAN that matches the DNS name contained within the **sub** value of its Entity Configuration.
4. **Certificate Revocation List (CRL)**: The participant MUST publish a CRL for its self-issued X.509 Certificates. This list MUST be accessible and regularly updated to ensure that any compromised or invalid X.509 Certificates are promptly revoked with the motivation of the revocation, if any.
5. **Basic Constraints**: The X.509 Certificate MUST include a ``Basic Constraints`` extension with ``CA:TRUE`` and a maximum path length of 1. This indicates that the Subordinate to which certificate is about, can only issue X.509 Certificates with a limited chain depth.
6. **Name Constraints**: The X.509 Certificate MUST include ``Name Constraints`` to specify permitted and excluded domains and URIs. For example:
   - Permitted:
     - ``URI.1=https://leaf.example.com``
     - ``DNS.1=leaf.example.com``
   - Excluded:
     - ``DNS=localhost``
     - ``DNS=localhost.localdomain``
     - ``DNS=127.0.0.1``
     - ``DNS=example.com``
     - ``DNS=example.org``
     - ``DNS=example.net``
     - ``DNS=*.example.org``
```

.. name-constraints::
   :permitted: URI.1=https://leaf.example.com
   :permitted: CN=leaf.example.com
   :permitted: DNS=leaf.example.com
   :excluded: DNS=localhost
   :excluded: DNS=localhost.localdomain
   :excluded: DNS=127.0.0.1
   :excluded: DNS=example.com
   :excluded: DNS=example.org
   :excluded: DNS=example.net
   :excluded: DNS=*.example.org

When a participant self-issues an X.509 Certificate, it adheres to the following requirements:

1. **Subject Name**: The X.509 Certificate's subject name MUST match the participant's identity. Specifically, the ``Common Name (CN)`` field should contain the Federation Entity unique identifier DNS name, which is included into the **sub** (subject) value in its federation Entity Configuration, removing ``https://`` and any webpaths.
2. **Subject Alternative Name (SAN)**: The X.509 Certificate MUST include a ``SAN URI`` that matches the **sub** value of its federation Entity Configuration.

> 3. **DNS Name**: The X.509 Certificate MUST include a DNS Name in the SAN that matches the DNS name contained within the **sub** value of its Entity Configuration.

Author: ```suggestion
3. **DNS Name**: The X.509 Certificate MUST include a DNS Name in the SAN that matches the DNS name contained within the **sub** value of its Entity Configuration, removing ``https://`` and any webpaths.
```

> 4. **Certificate Revocation List (CRL)**: The participant MUST publish a CRL for its self-issued X.509 Certificates. This list MUST be accessible and regularly updated to ensure that any compromised or invalid X.509 Certificates are promptly revoked with the motivation of the revocation, if any.

Author: ```suggestion
4. **Certificate Revocation List (CRL)**: If the issued X.509 Certificates has an expiration time superior to 24 hours, the X.509 Issuer MUST publish a CRL for the issued X.509 Certificates. This list MUST be accessible and regularly updated to ensure that any compromised or invalid X.509 Certificates are promptly revoked with the motivation of the revocation, if any.
```

> 5. **Basic Constraints**: The X.509 Certificate MUST include a ``Basic Constraints`` extension with ``CA:TRUE`` and a maximum path length of 1. This indicates that the Subordinate to which certificate is about, can only issue X.509 Certificates with a limited chain depth.

Author: ```suggestion
5. **Basic Constraints**: The X.509 Certificate MUST include a ``Basic Constraints`` extension with ``CA:TRUE`` and a maximum path length of 1 if the certificate issuer is a Federation Intermediate, if it is a Leaf, the maximum path length MUST be set to 0. This indicates that the Subordinate to which certificate is about, can only issue X.509 Certificates with a limited chain depth.
```

                Modulus:
                    00:af:82:3b:...
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Key Usage: 
                Digital Signature, Key Encipherment
            X509v3 Subject Alternative Name: 
                URI:https://leaf.example.org
            X509v3 Name Constraints: 
                Permitted:
                  URI.1=https://leaf.example.com
                  DNS.1=leaf.example.com
            X509v3 CRL Distribution Points: 
                Full Name:
                  URI:https://leaf.example.org/crl/self.crl

    Signature Algorithm: sha256WithRSAEncryption
         7d:6e:5f:...

Federation participants can ensure that their certificates are consistent, enabling interoperability and security across the federation. This approach, enabling X.509 certificate issuance delegation, introduces innovative practices for certificate management using the underlying trust relationships established within the OpenID Federation.

X.509 Certificate Revocation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

An X.509 Certificate can be revoked by its Issuer.

> When the X.509 Certificate issuer is the Leaf and therefore the X.509 Certificate is about itself, it MUST update its CRL.

Author: ```suggestion
Revocation lists, and or any other revocation check mechanisms, are required only for X.509 Certificate with expiration time superior to 24 hours, otherwise they are not required.

When the X.509 Certificate issuer is the Leaf and therefore the X.509 Certificate is about itself, if the certificate expiration time is superior than 24 hours from the ``X509_NOT_VALID_BEFORE`` time, it MUST implement a CRL about the issued certificate and keep it updated.
```

> When the X.509 Certificate issuer is an Immediate superior, such as the Trust Anchor or a Intermediate, it revokes the certificate about the leaf, therefore the X.509 Certificate about one of the Leaves Federation Entity Key. This action invalidates the entire Trust Chain associated with that Leaf's cryptographic public key, effectively removing its ability to issue further X.509 Certificates about itself. This hierarchical revocation mechanism ensures that any compromise or misbehavior by a Leaf entity can be swiftly addressed.

Author: ```suggestion
When the X.509 Certificate issuer is an Immediate superior, such as the Trust Anchor or a Intermediate, and it revokes the certificate about the Leaf, therefore the X.509 Certificate about one of the Leaves Federation Entity Key. This action invalidates the entire Trust Chain associated with that Leaf's cryptographic public key, effectively removing its ability to issue further X.509 Certificates about itself. This hierarchical revocation mechanism ensures that any compromise or misbehavior by a Leaf entity can be swiftly addressed.
```
